stream_reader.go

```diff
@@ -28,9 +30,56 @@ type streamReader[T streamable] struct {
 	errAccumulator utils.ErrorAccumulator
 	unmarshaler    utils.Unmarshaler
 
+	event       string
+	handlers    map[string]eventHandler[T]
+	handlerCtx  context.Context
+	handlerErr  error
+	lastRawLine []byte
+
 	httpHeader
 }
 
+type eventHandler[T streamable] func(T, []byte)
+
+func (stream *streamReader[T]) On(topic string, handler eventHandler[T]) error {
+	if len(topic) == 0 {
+		return ErrStreamEventEmptyTopic
+	}
+	if stream.handlers == nil {
+		stream.handlers = make(map[string]eventHandler[T])
+	}
+	stream.handlers[topic] = handler
+	return nil
+}
+
+func (stream *streamReader[T]) Run() error {
+	if stream.handlerCtx == nil {
+		stream.event = "message" // default event for chat completion stream
+		ctx, cancel := context.WithCancel(context.Background())
+		stream.handlerCtx = ctx
+		go func() {
```

Comment:
Sorry, I don't think we're going to merge this as starting goroutines don't fit into the design of this library.

Please refer to the following comments:
- https://github.com/sashabaranov/go-openai/pull/54#discussion_r1083519779
- https://github.com/sashabaranov/go-openai/issues/49#issuecomment-1373423407

Reply:
I'm happy to have this in the `examples/` folder, though!

Reply:
No problem. What if I revise the parts where I used goroutines? Would that work better with the design of the library?

Reply:
@CallOrRet I'm totally open to that!